airbyte-cdk/java/airbyte-cdk/s3-destinations/src/testFixtures/kotlin/io/airbyte/cdk/integrations/destination/s3/S3DestinationAcceptanceTest.kt

```diff
@@ -451,6 +451,66 @@ protected constructor(
         )
     }
 
+    /** Test runs 2 failed syncs and verifies the previous sync objects are not cleaned up. */
+    @Test
+    fun testOverwriteSyncMultipleFailedGenerationsFilesPreserved() {
+        assumeTrue(
+            implementsOverwrite(),
+            "Destination's spec.json does not support overwrite sync mode."
+        )
+        val config = getConfig()
+
+        // Run first failed attempt of same generation
+        val catalogPair =
+            getTestCatalog(SyncMode.FULL_REFRESH, DestinationSyncMode.OVERWRITE, 42, 12, 12)
+        val firstSyncMessages: List<AirbyteMessage> =
+            getFirstSyncMessagesFixture1(
+                catalogPair.first,
+                AirbyteStreamStatusTraceMessage.AirbyteStreamStatus.INCOMPLETE
+            )
+        try {
+            runSyncAndVerifyStateOutput(config, firstSyncMessages, catalogPair.first, false)
+            fail { "Should not succeed the sync when Trace message is INCOMPLETE" }
+        } catch (_: TestHarnessException) {}
+
+        // Run second failed attempt of same generation
+        val catalogPair2 =
+            getTestCatalog(SyncMode.FULL_REFRESH, DestinationSyncMode.OVERWRITE, 43, 12, 12)
+        val secondSyncMessages =
+            getFirstSyncMessagesFixture1(
+                catalogPair2.first,
+                AirbyteStreamStatusTraceMessage.AirbyteStreamStatus.INCOMPLETE
+            )
+
+        try {
+            runSyncAndVerifyStateOutput(config, secondSyncMessages, catalogPair2.first, false)
+            fail { "Should not succeed the sync when Trace message is INCOMPLETE" }
+        } catch (_: TestHarnessException) {}
+
+        // Verify our delayed delete logic creates no data downtime.
+        val defaultSchema = getDefaultSchema(config)
+        retrieveRawRecordsAndAssertSameMessages(
+            catalogPair.second,
+            firstSyncMessages + secondSyncMessages,
+            defaultSchema
+        )
+
+        // This doesn't happen in real world but just verifying if the generationId is incremented,
```

Comment:
fyi this is actually possible: if the user triggers a truncate refresh, it fails, they cancel it, and then run a _second_ truncate

Reply:
wait, really ? I thought genId is incremented only after a successful sync. I guess i misinterpreted. any terminal status COMPLETE/CANCELLED will result in this being incremented.

Reply:
it's not the terminal status, it's the user's action in the UI. If you click the refresh button, that triggers a gen ID increment

Reply:
Is that the only condition under which a refresh is triggered?

Reply:
Gen id goes up when
* A new scheduled overwrite sync starts?
* User clicks "sync now"?
* User makes a change and gets prompted to clear and start over?

What's the behavior for append syncs? Is each new append the same gen id? Gen id goes up only if they start over?

Reply:
gen ID increments when the user:
* runs a sync for a stream in full refresh mode (whether append or overwrite)
* clicks the "refresh" or "clear" button

For streams running incremental mode (whether append or dedup), normal syncs do _not_ increment gen ID - only refreshing/clearing the stream does that.